keps/sig-api-machinery/2885-server-side-unknown-field-validation/README.md

```diff
@@ -248,26 +257,72 @@ implementation. What is the desired outcome and how do we measure success?.
 The "Design Details" section below is for the real
 nitty-gritty.
 -->
-We propose using an opt-in API mechanism (such as content-type header or query
-param) to indicate to the server that it should fail when the kubernetes object
-in the request body supplied to POST, PUT, and PATCH requests contains
-extra/unknown fields.
 
-Clients such as kubectl will continue to use the `--validate=true` flag as
-before, but instead of triggering validation on the client-side, it will
-instruct the server to validate for unknown fields server-side.
+We propose adding server-side schema validation, enabling the API server
+to validate that objects received in the request body in POST, PUT, and
+PATCH requests contain no unknown or duplicate fields.
 
-This change will be made in at least two steps, one where we introduce the
-server side validation and a second where we modify kubectl to use the
-server-side validation (and mark the existing client-side validation as
-deprecated).
+We introduce a query parameter that enables the user/client to indicate
+that server-side validation should either not be performed at all, provide
+warnings for the presence of unknown/duplicate fields, or error out in the
+presence of unknown/duplicate fields.
+
+For beta, we propose modifying the kubectl `--validate` flag to enable users
+to choose between client-side validation or server-side
+validation (either strict or warning), when server-side validation is supported
+by the server (and default to strict server-side validation).
 
 ### Notes/Constraints/Caveats (Optional)
 
 #### Future Work
-After server side unknown field validation is implemented we can begin work to
-deprecate and remove client side validate and have kubectl use server side
-validation instead. A separate KEP will be published for that.
+
+##### Deprecating Client Side Validation
+Upon successfully providing kubectl access to server-side validation via the
+`--validate` flag, an open question will remain as to the future of client
+side validation.
+
+While we would like kubectl to use server-side validation by default once it
+is GA, the question remains whether client-side validation be optionally
+available to kubectl users or completely deprecated.
+
+Maintaining client-side validation indefinitely provides the benefit of
+allowing users to validate their objects without needing access to a server.
```

Comment:
> allowing users to validate their objects without needing access to a server

That ability is not present today, since the first step of client-side validation is fetching the schema from the server. Given the problems with client-side validation, I don't think we should expand use of it.

Reply:
I want to loop in @KnVerey and @seans3 from sig-cli, because when I brought up replacing CS validation with SS validation in kubectl initial, they expressed concern around this use case if we were to wholesale replace CS validation with SS validation without offering a CS alternative.

Correct me if I'm explaining this wrong, but I think the concern is that if a user has the schema cached (or gets it from somewhere other than the server), being able to validate my configs is still useful even if I don't have access to the server anymore (and is currently possible with existing CS validation), and we would be removing this functionality by removing access to CS validation

Reply:
client-side validation is known to be buggy, have false negatives, and (as exposed by kubectl) requires the schema which _usually_ requires server access. We *should* be removing it as soon as there's an equivalent server-side capability and we've passed the server/client skew window.

Reply:
(I agree on communicating/coordinating with sig-cli, but the current state of client-side validation in kubectl is not one we should preserve any longer than we have to)

Reply:
here's the source conversation for context https://youtu.be/OP84E63N81M?t=2798

Reply:
updated based on discussion at https://kubernetes.slack.com/archives/C2GL57FJ4/p1641572498004900

Reply:
Thanks for looping me in! FWIW after reading the conversation in Slack and reflecting on this further, I also agree with the intent-based flag values and the removal of the client-side implementation once the skew window has passed.